diff --git a/osdep/BSDEthernetTap.cpp b/osdep/BSDEthernetTap.cpp
index 8c6f94b..3d2f2d7 100644
--- a/osdep/BSDEthernetTap.cpp
+++ b/osdep/BSDEthernetTap.cpp
@@ -282,7 +282,19 @@ bool BSDEthernetTap::addIp(const InetAddress& ip)
 	else if (cpid > 0) {
 		int exitcode = -1;
 		::waitpid(cpid, &exitcode, 0);
-		return (exitcode == 0);
+		if (exitcode == 0) {
+			for (std::vector<InetAddress>::iterator i(_assignedIps.begin()); i != _assignedIps.end();) {
+				if (i->ipsEqual(ip))
+					i = _assignedIps.erase(i);
+				else
+					++i;
+			}
+			_assignedIps.push_back(ip);
+			std::sort(_assignedIps.begin(), _assignedIps.end());
+			_lastIfAddrsUpdate = 0;
+			return true;
+		}
+		return false;
 	}
 	return false;
 }
@@ -305,8 +317,16 @@ bool BSDEthernetTap::removeIp(const InetAddress& ip)
 		return false;
 	std::vector<InetAddress> allIps(ips());
 	if (std::find(allIps.begin(), allIps.end(), ip) != allIps.end()) {
-		if (___removeIp(_dev, ip))
+		if (___removeIp(_dev, ip)) {
+			for (std::vector<InetAddress>::iterator i(_assignedIps.begin()); i != _assignedIps.end();) {
+				if (i->ipsEqual(ip))
+					i = _assignedIps.erase(i);
+				else
+					++i;
+			}
+			_lastIfAddrsUpdate = 0;
 			return true;
+		}
 	}
 	return false;
 }
@@ -338,8 +358,13 @@ std::vector<InetAddress> BSDEthernetTap::ips() const
 	if (ifa)
 		freeifaddrs(ifa);
 
+	for (std::vector<InetAddress>::const_iterator i(_assignedIps.begin()); i != _assignedIps.end(); ++i) {
+		if (std::find(r.begin(), r.end(), *i) == r.end())
+			r.push_back(*i);
+	}
+
 	std::sort(r.begin(), r.end());
-	std::unique(r.begin(), r.end());
+	r.erase(std::unique(r.begin(), r.end()), r.end());
 
 	_ifaddrs = r;
 
@@ -394,7 +419,7 @@ void BSDEthernetTap::scanMulticastGroups(std::vector<MulticastGroup>& added, std
 		newGroups.push_back(MulticastGroup::deriveMulticastGroupForAddressResolution(*ip));
 
 	std::sort(newGroups.begin(), newGroups.end());
-	std::unique(newGroups.begin(), newGroups.end());
+	newGroups.erase(std::unique(newGroups.begin(), newGroups.end()), newGroups.end());
 
 	for (std::vector<MulticastGroup>::iterator m(newGroups.begin()); m != newGroups.end(); ++m) {
 		if (! std::binary_search(_multicastGroups.begin(), _multicastGroups.end(), *m))
diff --git a/osdep/BSDEthernetTap.hpp b/osdep/BSDEthernetTap.hpp
index 91c9a61..2022b16 100644
--- a/osdep/BSDEthernetTap.hpp
+++ b/osdep/BSDEthernetTap.hpp
@@ -72,6 +72,7 @@ class BSDEthernetTap : public EthernetTap {
 	volatile bool _enabled;
 	mutable std::vector<InetAddress> _ifaddrs;
 	mutable uint64_t _lastIfAddrsUpdate;
+	std::vector<InetAddress> _assignedIps;
 	std::vector<std::thread> _rxThreads;
 };
 
